### DISCUSSION TITLE
Using SetMemberAccessor to override object method

### DISCUSSION THREAD

**Discussion by luttje** [sebastienros/jint#949] [Q&A] [answered]
Hi y'all, loving Jint so far! Though I'm stuck atm.

I'm trying to make it so when I expose objects to JS, I can specify which members can be accessed from JS by adding a "[ScriptAccessible]" attribute. I can make this work for fields and properties, but am unsure how to approach returning a Method as a JsValue.

My setup:
```csharp
var engine = new Engine((_, opts) =>
{
    opts.SetMemberAccessor(GetObjectMember);
});

// [...]

// Hides anything that isn't tagged with [ScriptAccessible]
// Ensures lowerCamelCase is converted to UpperCamelCase
private static JsValue GetObjectMember(Engine engine, object target, string memberName)
{
    var builder = new StringBuilder();
    builder.Append(char.ToUpper(memberName[0]));

    if (memberName.Length > 1)
        builder.Append(memberName.Substring(1));

    var upperCamelCaseMemberName = builder.ToString();
    var type = target.GetType();
    var members = type.GetMember(upperCamelCaseMemberName);

    MemberInfo targetMember = null;

    for (var i = 0; i < members.Length; i++)
    {
        if (members[i].GetCustomAttribute<ScriptAccessibleAttribute>() != null)
            if (targetMember == null)
                targetMember = members[i];
            else
                // TODO: Implement checking overloaded method somehow
                throw new NotImplementedException($"Members name ambigous. Multiple members with name `{upperCamelCaseMemberName}` are marked [ScriptAccessible].");
    }

    // If no members exist marked [ScriptAccessible], do not allow accessing them
    if (targetMember == null)
        return JsValue.Undefined;

    switch (targetMember.MemberType)
    {
        case MemberTypes.Field:
            return JsValue.FromObject(engine, ((FieldInfo)targetMember).GetValue(target));
        case MemberTypes.Property:
            return JsValue.FromObject(engine, ((PropertyInfo)targetMember).GetValue(target));
        case MemberTypes.Method:
            return ???; // Return the method somehow?
        default:
            throw new NotImplementedException("Can't access this type of member from JS yet!");
    }
}
```

How can I return a method from C# such that it can be called in JavaScript?

I feel like the [classes and methods](https://github.com/sebastienros/jint/blob/af31a9dce60e805cc0675d29ffc6b77f23a489ce/Jint/Runtime/Interop/ObjectWrapper.cs#L117) I could use are marked internal or private so it seems I have to rewrite/copy a lot.

Anyone got an idea or pointers?

**Comment by lahma** [comment]
Things are usually internal and hidden to allow some room for change. In your case I think you're after a filter/predicate that could be called to verify whether member should be resolved?

I guess there's no such thing at the moment,  but there's been an ask to hide `GetType()` for example which currently allows caller to use reflection and thus bypass some security. So a filter like that would probably fit the bill for both use cases.

**Reply by luttje** [reply]
Yeah that sounds like what I'm looking for. The SetMemberAccessor is really a bit overkill for what I need. Indeed I'd just like to specify per member if it should be accessible from JS

**Reply by luttje** [reply]
I adjusted my question a bit, because I now think a filter or predicate is not really what I'm looking for. [This test ](https://github.com/sebastienros/jint/blob/20dcf69afd44c48601fb608ef7223413e7af66b9/Jint.Tests/Runtime/InteropTests.cs#L2282) already demonstrates how filtering could be done.

No I want to adjust which value or method is returned based on the request. In my use-case I want to maintain UpperCamelCase/PascalCase method and property names in C#, but be able to call those as lowerCamelCase from JavaScript

**Reply by luttje** [reply]
I'll keep trying. Perhaps providing a custom IReferenceResolver is the way to go.

I also just noticed that in js `someObject.test` by default resolves to the `Test` field on that object in C#. I wasn't aware of this feature.

**Comment by luttje** [comment, marked as answer]
I've found a way to return the method using ClrFunctionInstance. It was easier than it seemed though I'm not sure if there's a better way:
```csharp
return new ClrFunctionInstance(engine, memberName, (thisObject, arguments) =>
{
    return CallMethod(engine, method, methodParameters, (ObjectWrapper)thisObject, arguments);
}, methodParameters.Length, PropertyFlag.Configurable); // Why PropertyFlag.Configurable?? No idea.
```

I uploaded the full code to [this gist](https://gist.github.com/luttje/8266f5e972fa70dafc8cd2d79286cdbe) for future reference. Maybe it can help someone.

**Reply by lahma** [reply]
Great that you got that sorted and thanks for the provided sample. I think I'd like there to be an easier configuration story. So far what I've understood this is both filter (has attribute) and property name resolution (UpperCamelCase/IgnoreCase etc) configuration. So in this case implementing the member accessor is quite heavy-weight and also causes code duplication (which might also change in Jint when improvements/fixes are being made).

**Reply by luttje** [reply]
Thanks for your time. That sounds great. Curious to see which solution you will go for. Good luck

**Reply by lahma** [reply]
I think this should handle the filtering part https://github.com/sebastienros/jint/pull/951 .
